### DISCUSSION TITLE
Two factor default

### DISCUSSION THREAD

**Discussion by ghost** [OrchardCMS/OrchardCore#14617] [Q&A] [answered]
I have a requirement that all users use two factor auth.  Currently it's up to the individual user to enable two factor after they have their account.  Even if two factor is marked as required in the admin, it's still up to the individual user to go in and enable it for the requirement to apply to them.  Is there a way to enable two factor by default when adding a new user?  Will I have to replace the user module?

**Comment by hishamco** [comment]
There could be an option in the admin to make it compulsory for everyone to use 2FA or optional

I'm not sure if there's a reason why it doesn't support this

/cc @MikeAlhayek

**Comment by MikeAlhayek** [comment]
We have an option "Require two-factor authentication". If this is set, then when the user login, they will be required to enable two factor authentication.

you can also require-two factor authentication for a specific role only if you want.
![image](https://github.com/OrchardCMS/OrchardCore/assets/24724371/d080c265-db30-45cc-bb10-1d736b49be86)

**Reply by ghost** [reply]
With these boxes checked, any new user would still have to go into their accounts to enable 2fa.  If they don't do that, they won't be actually required to use 2fa.  We don't want new users to simply avoid enabling 2fa to skirt the requirement.

**Reply by ghost** [reply]
notice. "when selected any user 'with enabled two-factor..." we want to enable it for all users automatically at creation.

**Reply by MikeAlhayek** [reply]
The code seems to require it. if the new user logs in, they will be redirected to the page so they can enable 2FA. You can't manually enable. 2FA process requires them to confirm the second factor like scan the QR code or provide a validation code before we enable it.

**Reply by ghost** [reply]
It doesn't unless that individual enables 2fa after they have logged in.  See the sequence below.  I think it will demonstrate my point.  Slide 4 shows the new user login with no two factor even though it's set as require.  If the user enables 2fa on slide 5,  the requirement will actually go into effect

![adRequirement](https://github.com/OrchardCMS/OrchardCore/assets/96081566/268ece9f-d750-4abd-abc6-928ee1d2165f)

![addNewUser](https://github.com/OrchardCMS/OrchardCore/assets/96081566/e2aa9c61-9e8a-44a3-a185-86ba07e726de)

![newUserlogin](https://github.com/OrchardCMS/OrchardCore/assets/96081566/8c5e51ca-24a3-41b4-9916-8cb9f15e010b)

![loggedInno2fa](https://github.com/OrchardCMS/OrchardCore/assets/96081566/7d3ae4d1-7e0a-41cb-ab08-bb2b2441eed2)

![userHasToEnable2factor](https://github.com/OrchardCMS/OrchardCore/assets/96081566/0a3d460a-016b-4390-a72b-d0b0e6b20392)

**Reply by MikeAlhayek** [reply]
Are you able to by pass the Two-Factor authentication page without enabling the 2FA?

**Reply by ghost** [reply]
Yes.  It appears that my solution will be to create another user module that will enable 2fa after creation.  When the new user logs in for the 1st time they will get the app configuration page with no menu.  I'll enable the menu and other functionality only after they have configured the app.  Working my way through it now.

**Reply by ghost** [reply]
I selected the 2fa from security in the top menu (to get the last screenshot).  I could still edit content or whatever without setting up 2fa.

**Reply by MikeAlhayek** [reply]
That's not the expected behavior. You should not be able to bypass the 2FA page if it is required. I'll try to reproduce the issue.

**Reply by ghost** [reply]
I noticed that if you click "add", you are required to use 2fa on your next signin.  It doesn't matter if you enter a setup code or not.  So I altered the user creation as seen below.  I believe it will provide a workable "patch".

![code](https://github.com/OrchardCMS/OrchardCore/assets/96081566/f1eaaf2d-a41f-4ba3-a2bf-990ad09ff17b)
![screen 1](https://github.com/OrchardCMS/OrchardCore/assets/96081566/95a72453-401c-45f0-839e-a7fd872c6cf3)
![screen](https://github.com/OrchardCMS/OrchardCore/assets/96081566/ba15e6b3-5fad-4050-b426-7ae3c25cfe89)

**Reply by MikeAlhayek** [reply, marked as answer]
@tcobbfis indeed this is a bug in the code! I'll submit a fix. Meanwhile, if you want to fix this issue in your project register the following service

```
services.AddScoped<IUserClaimsProvider, TwoFactorAuthenticationClaimsProvider>();
```
